### DISCUSSION TITLE
Requirements for Lagoon to run on Managed Kubernetes Providers

### DISCUSSION THREAD

**Discussion by Schnitzel** [amazeeio/lagoon#2035] [General] [answered]
There are a couple of questions regarding using Lagoon on other Managed Kubernetes Providers, let's use this space to document requirements for other Managed Kubernetes Providers.

Read the original blog posts over at:
https://www.amazee.io/blog/post/why-were-moving-to-managed-kubernetes
https://www.amazee.io/blog/post/automating-migrations-between-lagoon-clusters-and-embracing-managed-kubernetes

**this is a WIP!**

**Comment by Schnitzel** [comment, marked as answer]
## Storage
Lagoon and it's deployed applications require the following type of storage types provided by Managed Kubernetes:

_The AWS, Azure, GCP Names are here only for reference purposes (so people understand what we mean with the types of storage). We want to make Lagoon able to run in any Managed Kubernetes Providers! Feel free to suggest any other providers!_

Type | k8s Access Modes | Used for | AWS, Azure, GCP Name  
------------ | ------------ | -------------  | -------------  
Block Storage | ReadWriteOnce | Lagoon: Docker Host Image Cache, Elasticsearch Data, API & Keycloak Database, etc <br> Applications: Solr Data, Persistent Redis, etc. | AWS EBS <br> Azure managed disks <br> GCP Persistent disks
Blob Storage (S3 Compatible) | ReadWriteOnce | Lagoon: Backup System <br> Applications: None | AWS S3 <br> Azure Blob storage (only via minio gateway) <br> GCP Cloud Storage
File Storage | ReadWriteMany | Lagoon: None <br> Applications: Persistent File Storage (like Drupal, Wordpress, etc.) | AWS EFS <br> Azure Files <br> GCP Filestore

If there is no File Storage available inside a managed Kubernetes the following alternatives could be used:
- **Deploy application pods always to a single node** <br> This removes the need for any File Storage type in the Kubernetes cluster, but also completely removes any resiliency against node-loss and also reduces the availability of the application. Is only suggested for development environments or environments that do not have any availability requirements. 
- **Deploy a ReadWriteMany storage into the cluster** <br> This is the most flexible way but also a complex one, as it requires the maintainers of the Kubernetes Cluster also to maintain a separate storage solution. There are many different ways on how to bring a ReadWriteMany storage into an k8s cluster, here a short list of them:  [OpenEBS](https://openebs.io/), [NFS Server](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs), External GlusterFS Cluster with [Gluster Provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/gluster)

**Comment by tobybellwood** [comment]
## Kubernetes Hosting Platform and Distribution

There are a number of managed Kubernetes hosting platforms out there that we could deploy Lagoon onto.  For us, the best way to ensure that the deployment is as smooth as possible and minimises rework to implement is to ensure that the platform meets the requirements of the Kubernetes Software Conformance Certification program.

All the hosting platforms listed on the [CNCF Landscape](https://landscape.cncf.io/category=certified-kubernetes-hosted&format=card-mode&grouping=category) have passed the necessary conformance tests for their supported releases of Kubernetes.  A new set of tests has to be passed for each release (1.15, 1.16, 1.17 etc) - and amazee.io will always prefer hosting with a certified conformant platform, utilising a supported release.

Additionally, the Kubernetes Software Conformance Certification program also allows for testing [Kubernetes Distributions](https://landscape.cncf.io/category=certified-kubernetes-distribution&format=card-mode&grouping=category).  Where we don't use a hosting platform (e.g. our current installs on Openshift Cloud Platform or in our local testing setup) we would utilise a certified distribution.
